Concurrent canister info calls without self-calls

AndyGura opened the pull request.

AndyGura committed:
implemented concurrent canister info calls without self-calls

AndyGura committed:
simplify logic

AndyGura committed:
adjustments

timohanke commented on src/history_be/CanisterHistory.mo:
```diff
@@ -95,14 +95,30 @@ module {
     /// If there is an ongoing sync, then new one is not starting.
     public func sync() : async* Bool {
-      if (sync_ongoing) return false;
+      let ?(ic, args) = schedule_sync_call() else return false;
```
Why don't we inline the function `schedule_sync_call()` here? It doesn't seem to do anything interesting.

timohanke commented on src/history_be/CanisterHistory.mo:
```diff
@@ -95,14 +95,30 @@ module {
+        ic,
+        {
+          canister_id;
+          num_requested_changes = ?Nat64.fromNat(20);
```
Maybe
```
num_requested_changes = ?(20 : Nat64);
```
or
```
num_requested_changes : ?Nat64 = ?20;
```

AndyGura replied:
Fixed

timohanke commented on src/history_be/main.mo:
```diff
@@ -109,17 +110,32 @@ actor class HistoryTracker() = self {
       let history = history_storage.get(index);
-      try { ignore async { await* history.sync() } } catch (_) {};
+      switch (history.schedule_sync_call()) {
+        case (?(ic, args)) Vec.add(calls, (history, ic.canister_info(args)));
```
I think we can remove `ic` from the return value of `schedule_sync_call`. Because here we get `ic` and `args` from `schedule_sync_call` but the function name, `canister_info` is hard-coded. That seems weird. We can hard-code `ic` as well. And only `args` comes from `schedule_sync_call`.

Or, alternatively, `schedule_sync_call` can return `async IC.CanisterInfoResponse` and then here we don't see any of ic, canister_info, args at all.

AndyGura replied:
The problem here is that `schedule_sync_call` is neither `async`nor `async*` function, so I cannot return futures from there, compiler returns error `send capability required, but not available`, even though I do not await it

AndyGura replied:
Fixed using your first proposal, hardcoding `ic`

timohanke commented on src/history_be/CanisterHistory.mo:
```diff
@@ -95,14 +95,30 @@ module {
-        canister_id;
-        num_requested_changes = ?Nat64.fromNat(20);
-      });
+    public func handle_sync_response(response : ?IC.CanisterInfoResponse) : Bool {
```
`handle_sync_response()` doesn't need a return value. Then `sync()` function can return true/false itself.

AndyGura replied:
Fixed

timohanke reviewed (commented):
Left some code comments

AndyGura committed:
Merge remote-tracking branch 'origin/develop' into feature/concurrency

AndyGura committed:
PR code adjustments

AndyGura committed:
PR code adjustments

AndyGura commented:
@timohanke somehow overseen your comments on this. Fixed now

timohanke committed:
Modularize concurrent calls logic

There is a module info/concurrent_calls.mo which contains all the logic of making concurrent calls, buffering the futures, awaiting the futures, catching errors, and detecting a trap in a continuation with a finally clause.

Because async generics do not exist in Motoko this module can not be a generic package. To make it as re-usable as possible it imports another module, info/single_call.mo, which exports the shared function that is the single call. It then wraps all the concurrency logic around the imported single calls.

The client code passes a list of call args in for all concurrent calls and a list of continuation callbacks, separated for the success case and the async error case.

There are also callbacks for each time a concurrent call is scheduled. Because it can happen that somewhere in the middle scheduling starts to fail because of a full queue. Through the callbacks the client code knows which ones have been scheduled.

The is also a single callback (not one for each concurrent call) for when a trap in a continuation is caught. It can normally only be used to shut down operations. As a rule, the continuation callbacks passed in by the client must not trap. If one of the callbacks traps then all subsequent futures will not be awaited anymore. They will be lost and the client code will not learn about there results. It will not even learn when or if they returned. The trap will be detected though and the trap callback called with a Nat value which is the number of calls that have been awaited so far and whose continuation did not trap.

The CanisterHistory class and main.mo code are adjusted to use the module.

timohanke committed:
Commit missed file

timohanke commented:
Superseded by (merged into) PR #16
